### PR TITLE
feat(app): create ODD protocol setup page

### PR DESCRIPTION
**Pull request by brenthagen**
# Overview

the first pass on the Protocol Setup page - the page hosts screens for instruments, modules, labware, etc and functional play and cancel buttons. The page is reached from the "Run protocol" button on the protocol details page, which creates a run. there are a few TODOs pending other pieces coming together. Components and styling are not final but close to what's currently in figma.

re RCORE-486

Screen Shot 2023-01-26 at 3 53 34 PM

Screen Shot 2023-01-26 at 3 51 00 PM

# Test Plan

 - smoke tested running protocols with and without modules on local emulated OT-3, OT-3 dev kit, OT-2 robot, and local simulated OT-2
 - added initial unit tests for Protocol Setup page, to be supplemented as functionality is added and finalized

# Changelog

 - Create ODD protocol setup page

# Review requests

 - look over general approach and structure of ProtocolSetup component
 - try out loading the setup page with a protocol or two and seeing if anything breaks - functionality isn't complete, but ideally we won't whitescreen. i've seen a couple 409s from stop requests.

# Risk assessment

low

### CODE REVIEW

File: app/src/App/OnDeviceDisplayApp.tsx [hunk 1]
```diff
@@ -100,15 +101,10 @@ export const onDeviceDisplayRoutes: RouteProps[] = [
...
-      <>
-        <BackButton />
-        <Box>protocol setup</Box>
-      </>
-    ),
+    Component: ProtocolSetup,
     exact: true,
     name: 'Protocol Setup',
-    path: '/protocols/:protocolId/:runId/setup',
+    path: '/protocols/:runId/setup',
```
Comment (shlokamin) [inline app/src/App/OnDeviceDisplayApp.tsx:93]: 😁

File: app/src/atoms/buttons/BackButton.tsx [hunk 2]
```diff
@@ -1,18 +1,24 @@
...
 import { useHistory } from 'react-router-dom'
 
 import { ALIGN_CENTER, Box, Btn, Flex, Icon } from '@opentrons/components'
 
-import { useTranslation } from 'react-i18next'
-
 // TODO(bh, 2022-12-7): finish styling when designs finalized
-export function BackButton(): JSX.Element {
+export function BackButton({
+  onClick,
```
Comment (shlokamin) [inline app/src/atoms/buttons/BackButton.tsx:9]: nice, i like this!

File: app/src/pages/OnDeviceDisplay/ProtocolSetup.tsx [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+  onClickSetupStep,
+  status,
+  title,
+  detail,
+  subDetail,
+}: ProtocolSetupStepProps): JSX.Element {
+  const backgroundColorByStepStatus = {
+    ready: 'rgba(4, 170, 101, 0.2)',
+    'not ready': '#FCF0D8',
+    general: '#E0E0E0',
```
Comment (shlokamin) [inline app/src/pages/OnDeviceDisplay/ProtocolSetup.tsx]: this is fine for now until design gets finalized, but any reason some of these are both hex + rgba?

File: app/src/pages/OnDeviceDisplay/ProtocolSetup.tsx [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+      <Flex
+        alignItems={ALIGN_CENTER}
+        backgroundColor={backgroundColorByStepStatus[status]}
+        borderRadius="1rem"
+        gridGap="1.5rem"
+        padding="1.5rem 1rem"
+      >
+        {status !== 'general' ? (
+          <Icon
+            color={status === 'ready' ? '#04AA65' : '#F09D20'}
```
Comment (shlokamin) [inline app/src/pages/OnDeviceDisplay/ProtocolSetup.tsx]: these colors will almost certainly change, but we've got these colors as style constants (`successEnabled`, and `warningEnabled`) if u wanna use em

File: app/src/pages/OnDeviceDisplay/ProtocolSetup.tsx [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+
+interface CloseButtonProps {
+  onClose: () => void
+}
+
+function CloseButton({ onClose }: CloseButtonProps): JSX.Element {
+  return (
+    <Btn
+      alignItems={ALIGN_CENTER}
+      border="2px solid #BF0000"
```
Comment (shlokamin) [inline app/src/pages/OnDeviceDisplay/ProtocolSetup.tsx]: same deal here with `errorEnabled`

File: app/src/pages/OnDeviceDisplay/ProtocolSetup.tsx [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+      borderRadius="4.25rem"
+      display={DISPLAY_FLEX}
+      height="4.25rem"
+      justifyContent={JUSTIFY_CENTER}
+      width="4.25rem"
+      onClick={onClose}
+      aria-label="close"
+    >
+      <Icon color="#BF0000" name="ot-close" size="2rem" />
```
Comment (shlokamin) [inline app/src/pages/OnDeviceDisplay/ProtocolSetup.tsx]: same deal here with `errorEnabled`

File: app/src/pages/OnDeviceDisplay/ProtocolSetup.tsx [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+interface PlayButtonProps {
+  disabled: boolean
+  onPlay: () => void
+}
+
+function PlayButton({ disabled, onPlay }: PlayButtonProps): JSX.Element {
+  return (
+    <Btn
+      alignItems={ALIGN_CENTER}
+      backgroundColor={disabled ? '#8F8F8F' : '#006CFA'}
```
Comment (shlokamin) [inline app/src/pages/OnDeviceDisplay/ProtocolSetup.tsx]: same thing here re: constants

File: app/src/pages/OnDeviceDisplay/ProtocolSetup.tsx [hunk 8]
```diff
@@ -0,0 +1,387 @@
...
+import type { OnDeviceRouteParams } from '../../App/types'
+
+interface ProtocolSetupStepProps {
+  onClickSetupStep: () => void
+  status: 'ready' | 'not ready' | 'general'
+  title: string
+  // first line of detail text
+  detail?: string
+  // second line of detail text
+  subDetail?: string
```
Comment (shlokamin) [inline app/src/pages/OnDeviceDisplay/ProtocolSetup.tsx]: might be worth changing these to accept `null` so you don't have to fallback to `undefined` in your ternaries

File: app/src/pages/OnDeviceDisplay/ProtocolSetup.tsx [hunk 9]
```diff
@@ -0,0 +1,387 @@
...
+      ? t('on-deck_labware', { count: onDeckLabwareCount })
+      : undefined
+  const labwareSubDetail =
+    additionalLabwareCount > 0
+      ? t('additional_labware', { count: additionalLabwareCount })
+      : undefined
+
+  return (
+    <>
+      {/* Protocol Setup Header */}
```
Comment (shlokamin) [inline app/src/pages/OnDeviceDisplay/ProtocolSetup.tsx:257]: more hex colors in this file that we've got in constants

**Review by shlokamin — APPROVED** [review]
code lgtm! left comments about using colors from our ui constants since some of the colors will probably stay the same. might be worth just using em now so we wont have to come back and convert them later.

sweet to see this coming together, great work. its very motivating to me, i hope it is to you too!

File: app/src/pages/OnDeviceDisplay/ProtocolSetup.tsx (hunk 3 above)
Comment (brenthagen) [reply app/src/pages/OnDeviceDisplay/ProtocolSetup.tsx]: just transcribing directly from figma. may have already changed

**Pull request merged**